enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+
+##### Examples
+
+```json
+"runtimeInfo": {
+  "os": "yqx1WBefoIAq",
+  "osVersion": "AQN-qh_OyRUA",
+  "kind": "wbpgzhNYZQOi",
+  "kindVersion": "cAsoUDy8JFNV",
+  "kindImplementer": "DzP5-GS6tDOQ",
+  "runtimes": [
+    {
+      "name": "zBVuVhrrC_vC",
+      "version": "lyAE7Oh63WPU"
+    }
+  ]
+}
+```
+
+As the values are derived by the insights-operator-runtime component or correspond to software versions, it is possible to create a dictionary of the hashes and retrieve the unobfuscated values. The example above corresponds to:
+
+```json
+"runtimeInfo": {
+  "os": "rhel",
+  "osVersion": "8.9",
+  "kind": "Java",
+  "kindVersion": "17.0.7",
+  "kindImplementer": "Red Hat, Inc.",
+  "runtimes": [
+    {
+      "name": "Quarkus",
+      "version": "2.13.8.Final-redhat-00004"
+    }
+  ]
+}
+```
+
+### Workflow Description
+
+In its existing implementation, the OpenShift Insights Operator periodically gathers workload data every 12 hours.
+
+The Operator iterates over pods in the cluster (up to a limit of 8000) and, for each pod's containers, captures their Image ID and entry points.
+
+When the operator starts gathering workload information, it would first gather runtime information for all the containers running on the cluster's worker nodes and then
+populate the existing gathered info with these additional runtime information.
+
+![insights-operator-runtime workflow](./insights-operator-runtime-1.png "Insights Operator Runtime Worklow")
+
+This is not a user-facing feature. It is achieved within the OpenShift Insights Operator gathering workflow.
+The user is able to examine the gathered data by unarchiving the archive generated by the operator.
+
+The DaemonSet for the insights-operator-runtime is deployed when the Insights Operator is installed in the cluster and is not accessible outside of the `openshift-insights` namespace.
+
+From the operator perspective, the insights-operator-runtime is a black box, and its requirements are:
+
+* Determine the Pod name of each insights-operator-runtime’s Pod (as the insights-operator-runtime is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.
+  * There is no service exposing the insights-operator-runtime and its pods are accessed directly.
+* When the operator starts gathering workload information, execute the  `/scan-containers` executable on each insights-operator-runtime pod. The executable will return a JSON payload in the standard output that represents a tree of `namespace/pod-name/container-id/runtime-info`.
+* As the operator iterates over the containers in the cluster, check if there is a runtime-info entry for the `namespace/pod-name/container-id` and add it to the existing `workloadContainerShape` (after hashing all the values)
+
+The insights-operator-runtime functionality is provided by an image that must be pullable from the `openshift-insights` namespace and present in the cluster at installation.
```

Comment:
See [these docs][1] about adding new images to OpenShift releases.  It might be easier to build this new component into [the insights operator repository][2] as an alternative command (the way the machine-config operator does][3]), so you can run from that existing OpenShift-payload image.

[1]: https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#product-builds-and-becoming-part-of-an-openshift-release
[2]: https://github.com/openshift/insights-operator/ 
[3]: https://github.com/openshift/machine-config-operator/tree/612b9cd6a642a560bb48b9e8f1fa9cb4b54622ab/cmd

Reply:
I guess the later:

> In order to minimize any security risks, its code will be audited for security and its image will be as minimal as possible.

excludes building this from the Insights Operator repository, but you'll still want to go through [the OCP-image-onboarding doc][1] and... wait, how are we saying that the binaries in the runtime image are going to be so secure that even if a not-as-closely-audited Insights operator is running whatever it wants inside those containers, it will still be sufficiently secure?

[1]: https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#product-builds-and-becoming-part-of-an-openshift-release